Products/ZenHub/invalidationmanager.py

```diff
@@ -0,0 +1,270 @@
+                filters, key=lambda f: getattr(f, 'weight', 100)
+            ):
+                fltr.initialize(self.__dmd)
+                self._invalidation_filters.append(fltr)
+            self.log.debug('Registered %s invalidation filters.',
+                           len(self._invalidation_filters))
+            return self._invalidation_filters
+        except Exception:
+            log.exception('error in initialize_invalidation_filters')
+
+    @inlineCallbacks
+    def process_invalidations(self):
+        '''Periodically process database changes.
+        synchronize with the database, and poll invalidated oids from it,
+        filter the oids,  send them to the invalidation_processor
+
+        @return: None
+        '''
+        try:
+            now = time()
+            yield self._syncdb()
+            if self._paused():
+                return
+
+            oids = self._poll_invalidations()
+            if not oids:
+                log.debug('no invalidations found: oids=%s' % oids)
+                return
+
+            for oid in oids:
+                yield self.invalidation_pipeline.run(oid)
+
+            self.log.debug('Processed %s raw invalidations', len(oids))
+
+        except Exception:
+            log.exception('error in process_invalidations')
+        finally:
+            self.totalEvents += 1
+            self.totalTime += time() - now
+            log.debug('end process_invalidations')
+
+    @inlineCallbacks
+    def _syncdb(self):
+        try:
+            self.log.debug("[processQueue] syncing....")
+            yield self.__syncdb()
+            self.log.debug("[processQueue] synced")
+        except Exception as err:
+            self.log.warn("Unable to poll invalidations, will try again.")
+
+    def _paused(self):
+        if not self._currently_paused:
+            if self.__dmd.pauseHubNotifications:
+                self._currently_paused = True
+                log.info('notifications have been paused')
+                self._send_event(self._invalidation_paused_event)
+                return True
+            else:
+                return False
+
+        else:
+            if self.__dmd.pauseHubNotifications:
+                log.debug('notifications are paused')
+                return True
+            else:
+                self._currently_paused = False
+                log.info('notifications unpaused')
+                self._send_event(self._invalidation_unpaused_event)
+                return False
+
+    def _poll_invalidations(self):
+        '''pull a list of invalidated object oids from the database
+        '''
+        try:
+            log.debug('poll invalidations from dmd.storage')
+            return self.__poll_invalidations()
+        except Exception:
+            log.exception('error in _poll_invalidations')
+
+    @inlineCallbacks
+    def _send_event(self, event):
+        yield self.__send_event(event)
+
+
+class InvalidationPipeline(object):
+
```

Comment:
I think this class could use a docstring to explain what it does.

Reply:
Added doc string